crates/core/src/subscription/module_subscription_actor.rs

```diff
@@ -73,35 +77,52 @@ impl ModuleSubscriptionManager {
             .map_err(|_| NoSuchModule)
     }
 
-    pub async fn broadcast_event(&self, client: Option<&ClientConnectionSender>, mut event: ModuleEvent) {
+    pub async fn broadcast_event(&self, client: Option<&ClientConnectionSender>, event: ModuleEvent) {
         match event.status {
             EventStatus::Committed(_) => {
                 self.tx
                     .send(Command::BroadcastCommitEvent { event })
                     .expect("subscription actor panicked");
             }
             EventStatus::Failed(_) => {
-                if let Some(client) = client {
-                    let message = TransactionUpdateMessage {
-                        event: &mut event,
-                        database_update: Default::default(),
-                    };
-                    let _ = client.send_message(message).await;
-                } else {
-                    log::trace!("Reducer failed but there is no client to send the failure to!")
-                }
+                self.handle_failed_event(client, event).await;
             }
             EventStatus::OutOfEnergy => {} // ?
         }
     }
 
     pub fn broadcast_event_blocking(&self, client: Option<&ClientConnectionSender>, event: ModuleEvent) {
-        tokio::runtime::Handle::current().block_on(self.broadcast_event(client, event))
+        tokio::runtime::Handle::current().block_on(async move {
+            match event.status {
+                EventStatus::Committed(_) => {
+                    if let Err(e) = self.actor.lock_arc().broadcast_commit_event(event).await {
```

Comment:
Won't this lock be released once the message has been received on the other end? Don't we want to hold the lock until all of the subscriptions have been evaluated?

Reply:
thanks to point, I would live to hold this PR a bit, I think this can be done in much better way directly in subscription actor instead of here, so moved it to draft.